[WK2] ArgumentCoder<std::variant<...>> should use increasing index sequences

zdobersek opened the pull request:
#### 68fb45e34eda2fd008bee2558273e6b61f21a99f
<pre>
[WK2] ArgumentCoder&lt;std::variant&lt;...&gt;&gt; should use increasing index sequences
<a href="https://bugs.webkit.org/show_bug.cgi?id=249525">https://bugs.webkit.org/show_bug.cgi?id=249525</a>

Reviewed by Kimmo Kinnunen.

More intuitive increasing index sequences are adopted in the encoding
and decoding implementations of ArgumentCoder&lt;std::variant&lt;...&gt;&gt;, with
no real change in behavior.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:

Canonical link: <a href="https://commits.webkit.org/258198@main">https://commits.webkit.org/258198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cfd17498c48dc015c69266c4d336f6221ac0df05

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34255 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110485 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11291 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1222 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/93602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108317 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106983 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/93602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/23230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/93602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3996 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/4044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1164 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10147 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/44231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5632 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5797 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -705,21 +705,20 @@ template<typename... Types> struct ArgumentCoder<std::variant<Types...>> {
-        if (Index == i) {
-            encoder << std::get<Index>(std::forward<T>(variant));
-            return;
+        constexpr size_t Index = sizeof...(Indices);
```
IIRC this was discussed also last time: this variable could be just `index`, like other constexpr variables.

zdobersek replied:
Yeah, will change it.

kkinnunen-apple reviewed (approved):
Not common for a template change: it is simpler!

zdobersek committed:
[WK2] ArgumentCoder<std::variant<...>> should use increasing index sequences
https://bugs.webkit.org/show_bug.cgi?id=249525

Reviewed by Kimmo Kinnunen.

More intuitive increasing index sequences are adopted in the encoding
and decoding implementations of ArgumentCoder<std::variant<...>>, with
no real change in behavior.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:

Canonical link: https://commits.webkit.org/258198@main
